src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,153 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> first;
+    private Node<T> last;
+    private int size;
+
+    private static class Node<T> {
+        private Node<T> prev;
+        private Node<T> next;
+        private T value;
+
+        public Node(Node<T> prev, T value, Node<T> next) {
+            this.value = value;
+            this.prev = prev;
+            this.next = next;
+        }
+    }
+
     @Override
     public void add(T value) {
+        Node<T> addedElement = new Node<>(null, value, null);
+        if (isEmpty()) {
+            first = addedElement;
+            last = addedElement;
+        } else {
+            addedElement.prev = last;
+            last.next = addedElement;
+            last = addedElement;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index > size() || index < 0) {
+            throw new IndexOutOfBoundsException("Index " + index
+                   + " out of List bounds with size " + size);
+        }
+        if (index == size()) {
+            add(value);
+            return;
+        }
+        Node<T> nextElement = findElement(index);
+        Node<T> addedElement = new Node<>(nextElement.prev, value, nextElement);
+        if (nextElement.prev != null) {
+            nextElement.prev.next = addedElement;
+        } else {
+            first = addedElement;
+        }
+        nextElement.prev = addedElement;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element: list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> currentElement = findElement(index);
+        return currentElement.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> currentElement = findElement(index);
+        T changedValue = currentElement.value;
+        currentElement.value = value;
+        return changedValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        T deletedValue;
+        if (size == 1) {
+            size = 0;
+            return first.value;
+        }
+        if (index == size - 1) {
+            deletedValue = last.value;
+            last = last.prev;
+            last.next = null;
+            size--;
+            return deletedValue;
+        }
+        if (index == 0) {
+            deletedValue = first.value;
+            first = first.next;
+            first.prev = null;
+            size--;
+            return deletedValue;
+        }
```

Comment:
better move this type of logic to unlink. Because it is still logic to unlink node